### PR TITLE
Added Material Python interface

### PR DESCRIPTION
**Pull request by ahcorde**
Signed-off-by: ahcorde <ahcorde@gmail.com>

# 🎉 New feature

This PR is part of this meta ticket https://github.com/ignitionrobotics/sdformat/issues/931

## Summary

Added Material Python interface

## Test it

```python
from sdformat import Material

material = Material()
```

## Checklist
- [x] Signed all commits for DCO
- [x] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [x] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [x] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [x] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

### CODE REVIEW

File: python/src/sdf/pyMaterial.cc [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+  pybind11::class_<sdf::Material> materialModule(module, "Material");
+  materialModule
+    .def(pybind11::init<>())
+    .def(pybind11::init<sdf::Material>())
+    .def("ambient", &sdf::Material::Ambient,
+         "Get the ambient color. The ambient color is "
+         "specified by a set of three numbers representing red/green/blue, "
+         "each in the range of [0,1].")
+    .def("set_ambient", &sdf::Material::SetAmbient,
+         "Get the ambient color. The ambient color is "
```
Comment (azeey) [inline python/src/sdf/pyMaterial.cc:44]: nit: Get->Set

File: python/src/sdf/pyMaterial.cc [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+         "specified by a set of three numbers representing red/green/blue, "
+         "each in the range of [0,1].")
+    .def("diffuse", &sdf::Material::Diffuse,
+         "Get the diffuse color. The ambient color is "
+         "specified by a set of three numbers representing red/green/blue, "
+         "each in the range of [0,1].")
+    .def("set_diffuse", &sdf::Material::SetDiffuse,
+         "Get the diffuse color. The ambient color is "
```
Comment (azeey) [inline python/src/sdf/pyMaterial.cc:52]:
```suggestion
         "Set the diffuse color. The diffuse color is "
```

File: python/src/sdf/pyMaterial.cc [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+    .def("ambient", &sdf::Material::Ambient,
+         "Get the ambient color. The ambient color is "
+         "specified by a set of three numbers representing red/green/blue, "
+         "each in the range of [0,1].")
+    .def("set_ambient", &sdf::Material::SetAmbient,
+         "Get the ambient color. The ambient color is "
+         "specified by a set of three numbers representing red/green/blue, "
+         "each in the range of [0,1].")
+    .def("diffuse", &sdf::Material::Diffuse,
+         "Get the diffuse color. The ambient color is "
```
Comment (azeey) [inline python/src/sdf/pyMaterial.cc:48]:
```suggestion
         "Get the diffuse color. The diffuse color is "
```

File: python/src/sdf/pyMaterial.cc [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+         "specified by a set of three numbers representing red/green/blue, "
+         "each in the range of [0,1].")
+    .def("set_diffuse", &sdf::Material::SetDiffuse,
+         "Get the diffuse color. The ambient color is "
+         "specified by a set of three numbers representing red/green/blue, "
+         "each in the range of [0,1].")
+    .def("specular", &sdf::Material::Specular,
+         "Get the specular color. The ambient color is "
```
Comment (azeey) [inline python/src/sdf/pyMaterial.cc:56]:
```suggestion
         "Get the specular color. The specular color is "
```

File: python/src/sdf/pyMaterial.cc [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+         "specified by a set of three numbers representing red/green/blue, "
+         "each in the range of [0,1].")
+    .def("set_specular", &sdf::Material::SetSpecular,
+         "Get the specular color. The ambient color is "
```
Comment (azeey) [inline python/src/sdf/pyMaterial.cc:60]:
```suggestion
         "Set the specular color. The specular color is "
```

File: python/src/sdf/pyMaterial.cc [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+         "specified by a set of three numbers representing red/green/blue, "
+         "each in the range of [0,1].")
+    .def("emissive", &sdf::Material::Emissive,
+         "Get the emissive color. The ambient color is "
```
Comment (azeey) [inline python/src/sdf/pyMaterial.cc:64]:
```suggestion
         "Get the emissive color. The emissive color is "
```

File: python/src/sdf/pyMaterial.cc [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+         "specified by a set of three numbers representing red/green/blue, "
+         "each in the range of [0,1].")
+    .def("set_emissive", &sdf::Material::SetEmissive,
+         "Get the emissive color. The ambient color is "
```
Comment (azeey) [inline python/src/sdf/pyMaterial.cc:68]:
```suggestion
         "Set the emissive color. The emissive color is "
```

File: python/test/pyMaterial_TEST.py [hunk 8]
```diff
@@ -0,0 +1,244 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import copy
+from sdformat import Material
+from ignition.math import Color
+import unittest
+
+
+class GeometryTEST(unittest.TestCase):
```
Comment (azeey) [inline python/test/pyMaterial_TEST.py:21]:
```suggestion
class MaterialTEST(unittest.TestCase):
```

File: python/test/pyMaterial_TEST.py [hunk 9]
```diff
@@ -0,0 +1,244 @@
...
+    self.assertEqual(5, moved.render_order())
+    self.assertTrue(moved.double_sided())
+    self.assertEqual("uri", moved.script_uri())
+    self.assertEqual("name", moved.script_name())
+    self.assertEqual(Material.ShaderType.VERTEX, moved.shader())
+    self.assertEqual("map", moved.normal_map())
+    # TODO(ahcorde) Add Pbr python interface
+    # self.assertEqual(workflow,
+    #   *moved.PbrMaterial()->Workflow(sdf::PbrWorkflowType::METAL))
+    self.assertEqual("/my/path", moved.file_path())
```
Comment (azeey) [inline python/test/pyMaterial_TEST.py:240]: These checks can be removed since there no move assignment in python.

**Review by azeey — APPROVED** [review]

File: python/src/sdf/pyMaterial.cc (hunk 1 above)
Comment (azeey) [reply python/src/sdf/pyMaterial.cc]: https://github.com/ignitionrobotics/sdformat/pull/948/commits/7b42ea9b03e17d2be4d5979326f0322af659961d

File: python/src/sdf/pyMaterial.cc (hunk 2 above)
Comment (azeey) [reply python/src/sdf/pyMaterial.cc]: https://github.com/ignitionrobotics/sdformat/pull/948/commits/7b42ea9b03e17d2be4d5979326f0322af659961d

File: python/src/sdf/pyMaterial.cc (hunk 3 above)
Comment (azeey) [reply python/src/sdf/pyMaterial.cc]: https://github.com/ignitionrobotics/sdformat/pull/948/commits/7b42ea9b03e17d2be4d5979326f0322af659961d

File: python/src/sdf/pyMaterial.cc (hunk 5 above)
Comment (azeey) [reply python/src/sdf/pyMaterial.cc]: https://github.com/ignitionrobotics/sdformat/pull/948/commits/7b42ea9b03e17d2be4d5979326f0322af659961d

File: python/src/sdf/pyMaterial.cc (hunk 7 above)
Comment (azeey) [reply python/src/sdf/pyMaterial.cc]: https://github.com/ignitionrobotics/sdformat/pull/948/commits/7b42ea9b03e17d2be4d5979326f0322af659961d

File: python/test/pyMaterial_TEST.py (hunk 8 above)
Comment (azeey) [reply python/test/pyMaterial_TEST.py]: https://github.com/ignitionrobotics/sdformat/pull/948/commits/7b42ea9b03e17d2be4d5979326f0322af659961d

File: python/test/pyMaterial_TEST.py (hunk 9 above)
Comment (azeey) [reply python/test/pyMaterial_TEST.py]: https://github.com/ignitionrobotics/sdformat/pull/948/commits/7b42ea9b03e17d2be4d5979326f0322af659961d

Also added a test for deepcopy in https://github.com/ignitionrobotics/sdformat/pull/948/commits/d9a68bf6a41afee265ebb1d9bfc762352312bd45

File: python/src/sdf/pyMaterial.cc [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+    .def("set_file_path", &sdf::Material::SetFilePath,
+         "Set the path to the file where this element was loaded from.")
+    .def("__copy__", [](const sdf::Material &self) {
+      return sdf::Material(self);
+    })
+    .def("__deepcopy__", [](const sdf::Material &self, pybind11::dict) {
+      return sdf::Material(self);
+    }, "memo"_a);
+
+    pybind11::enum_<sdf::ShaderType>(materialModule, "ShaderType")
```
Comment (azeey) [inline python/src/sdf/pyMaterial.cc:118]: This enum is in the `sdf` namespace in C++, but under `sdformat.Material` in python. Should we make them consistent?

**Comment by ahcorde** [comment]
@osrf-jenkins retest this please

**Pull request merged**